src/main/java/com/scalar/db/storage/common/checker/OperationChecker.java

```diff
@@ -0,0 +1,281 @@
+import com.scalar.db.storage.common.util.Utility;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public final class OperationChecker {
+  private OperationChecker() {}
+
+  public static void check(Get get, TableMetadata metadata) {
+    checkProjections(get, metadata);
+
+    if (Utility.isSecondaryIndexSpecified(get, metadata)) {
+      if (!new ColumnChecker(metadata).check(get.getPartitionKey().get().get(0))) {
+        throw new IllegalArgumentException(
+            "The partition key is not properly specified. Operation: " + get);
+      }
+
+      if (get.getClusteringKey().isPresent()) {
+        throw new IllegalArgumentException(
+            "Clustering keys cannot be specified when using an index. Operation: " + get);
+      }
+      return;
+    }
+
+    checkPrimaryKey(get, metadata);
+  }
+
+  public static void check(Scan scan, TableMetadata metadata) {
+    checkProjections(scan, metadata);
+
+    if (Utility.isSecondaryIndexSpecified(scan, metadata)) {
+      if (!new ColumnChecker(metadata).check(scan.getPartitionKey().get().get(0))) {
+        throw new IllegalArgumentException(
+            "The partition key is not properly specified. Operation: " + scan);
+      }
+
+      if (scan.getStartClusteringKey().isPresent() || scan.getEndClusteringKey().isPresent()) {
+        throw new IllegalArgumentException(
+            "Clustering keys cannot be specified when using an index. Operation: " + scan);
+      }
+      if (!scan.getOrderings().isEmpty()) {
+        throw new IllegalArgumentException(
+            "Orderings cannot be specified when using an index. Operation: " + scan);
+      }
+      return;
+    }
+
+    checkPartitionKey(scan, metadata);
+
+    checkClusteringKeys(scan, metadata);
+
+    if (scan.getLimit() < 0) {
+      throw new IllegalArgumentException("The limit cannot be negative Operation: " + scan);
+    }
+
+    checkOrderings(scan, metadata);
+  }
+
+  private static void checkProjections(Selection selection, TableMetadata tableMetadata) {
+    for (String projection : selection.getProjections()) {
+      if (!tableMetadata.getColumnNames().contains(projection)) {
+        throw new IllegalArgumentException(
+            "The specified projection is not found. Operation: " + selection);
+      }
+    }
+  }
+
+  private static void checkClusteringKeys(Scan scan, TableMetadata metadata) {
+    scan.getStartClusteringKey()
+        .ifPresent(startClusteringKey -> checkStartClusteringKey(scan, metadata));
+
+    scan.getEndClusteringKey().ifPresent(endClusteringKey -> checkEndClusteringKey(scan, metadata));
+
+    if (scan.getStartClusteringKey().isPresent() && scan.getEndClusteringKey().isPresent()) {
+      Key startClusteringKey = scan.getStartClusteringKey().get();
+      Key endClusteringKey = scan.getEndClusteringKey().get();
+      Supplier<String> message =
```

Comment:
Just a question.
Is there any benefit of doing this instead of just doing `String message = " ..."`?

Reply:
Here, I wanted to do a lazy evaluation for the plus (+) operation in building the message. If just doing `String message = "..."`, the plus operation is executed even if `message` is not used with IllegalArgumentException. That's why I use `Supplier` here.

Reply:
Good to know! Thanks!